From: Jay Hernandez <jay@chelsio.com>
Date: Thu, 30 May 2013 03:24:14 +0000
Subject: [PATCH 048/223] cxgb4: Force uninitialized state if FW_ON_ADAPTER is
 < FW_VERSION and we're the MASTER_PF
Patch-mainline: v3.11-rc1
Git-commit: e69972f52bdc4bc384a21a6561d5673b4a94e989 (partial)
References: bsc#909577 FATE#317550

Forcing uninitialized state allows us to upgrade and reinitialize the adapter.

FW_VERSION_T4 = 1.4.0.0
FW_VERSION_T5 = 0.0.0.0
At this point driver supports above and greater than above version of firmware.
If it doesn't find the required firmware version than it forces the adapter to
be reinitialized as shown below.

1) If FW_ON_ADAPTER < FW_VERSION and we're the MASTER_PF force uninitialized
   state and a FW upgrade if available.

       - If FW_ON_ADAPTER < /lib/firmware/cxgb4/t*fw.bin we will update the
         adapters FW.
       - If FW_ON_ADAPTER >= /lib/firmware/cxgb4/t*fw.bin don't upgrade FW.
       - If upgrade_fw() fails force reinitialization of the adapter anyways,
         it might still work.

   Either way forcing the uninitialized state allows cxgb4 reinitialize FW.

2) If FW_ON_ADAPTER >= FW_VERSION driver follows normal path.

Signed-off-by: Jay Hernandez <jay@chelsio.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: David Chang <dchang@suse.com>
---
 drivers/net/ethernet/chelsio/cxgb4/t4_hw.c |   11 ++++-------
 1 file changed, 4 insertions(+), 7 deletions(-)

--- a/drivers/net/ethernet/chelsio/cxgb4/t4_hw.c
+++ b/drivers/net/ethernet/chelsio/cxgb4/t4_hw.c
@@ -938,15 +938,12 @@ int t4_check_fw_version(struct adapter *
 	memcpy(adapter->params.api_vers, api_vers,
 	       sizeof(adapter->params.api_vers));
 
-	if (major < FW_VERSION_MAJOR ||
-	    (major == FW_VERSION_MAJOR && minor < FW_VERSION_MINOR) ||
-	    (major == FW_VERSION_MAJOR && minor == FW_VERSION_MINOR &&
-	     micro < FW_VERSION_MICRO)) {
+	if (major < exp_major || (major == exp_major && minor < exp_minor) ||
+	    (major == exp_major && minor == exp_minor && micro < exp_micro)) {
 		dev_err(adapter->pdev_dev,
-			"Card has fimrware version %u.%u.%u, minimum "
+			"Card has firmware version %u.%u.%u, minimum "
 			"supported firmware is %u.%u.%u.\n", major, minor,
-			micro, FW_VERSION_MAJOR, FW_VERSION_MINOR,
-			FW_VERSION_MICRO);
+			micro, exp_major, exp_minor, exp_micro);
 		return -EFAULT;
 	}
 
